### ISSUE TITLE
Difference in output for BART when beam size > 1

### ISSUE THREAD

**Issue by shrey-singla** [OpenNMT/CTranslate2#1493]
I converted the model checkpointed as "knkarthick/bart-large-xsum-samsum" on huggingface using the command 
`ct2-transformers-converter --model "knkarthick/bart-large-xsum-samsum" --output_dir "bart-large-xsum-samsum-ct"`

Then I ran the script: 

```
import pandas as pd
import torch

from transformers import BartForConditionalGeneration, AutoTokenizer
from transformers import AutoTokenizer
import ctranslate2


device_type = 'cuda'
device_id = 6
device_str = f'{device_type}:{device_id}'
device = torch.device(device_str)

model_name = "knkarthick/bart-large-xsum-samsum"
model = BartForConditionalGeneration.from_pretrained(model_name).eval().to(device)
tokenizer = AutoTokenizer.from_pretrained(model_name)



fast_model =  ctranslate2.Translator("bart-large-xsum-samsum-ct",
                                     device=device_type,
                                     device_index=device_id)



text = ["""
Hannah: Hey, do you have Betty's number?
Amanda: Lemme check
Amanda: Sorry, can't find it.
Amanda: Ask Larry
Amanda: He called her last time we were at the park together
Hannah: I don't know him well
Amanda: Don't be shy, he's very nice
Hannah: If you say so..
Hannah: I'd rather you texted him
Amanda: Just text him 🙂
Hannah: Urgh.. Alright
Hannah: Bye
Amanda: Bye bye
"""]


def get_out(inp, model):
    inputs = tokenizer(inp, max_length=1024, truncation=True, return_tensors="pt", padding="max_length")
    summary_ids = model.generate(inputs["input_ids"].to(device), 
                                num_beams=3,
                                min_length=20, 
                                max_length=1024,
                                )
    chars = tokenizer.batch_decode(summary_ids, skip_special_tokens=True, clean_up_tokenization_spaces=False)[0]
    return  chars


def get_out_fast(inp, model):
    source = tokenizer.encode(inp[0])
    source = tokenizer.convert_ids_to_tokens(source)
    results = model.translate_batch([source], 
                                     beam_size=3,
                                     min_decoding_length=20,
                                     max_decoding_length=1024,
                                     no_repeat_ngram_size=3,
                                     )
    target = results[0].hypotheses[0]
    return tokenizer.decode(tokenizer.convert_tokens_to_ids(target), skip_special_tokens=True)

res_vanilla = get_out(text, model)
res_fast = get_out_fast(text, fast_model)

print("Vanilla output: \n", res_vanilla)
print("Ctranslate output: \n", res_fast)

```

Outputs:
```
Vanilla output: 
 Amanda can't find Betty's number. She wants to ask Larry to text him.
Ctranslate output: 
 Amanda can't find Betty's number. She can't remember where she gave it to her last time they were at the park together. She wants to text Larry.
```

I expect the output to be exactly same. However, I observe as shown above that the ctranslate2 model changes the output. I observed this behaviour only when the search algorithm is beam search. (In the greedy search, the outputs match at the logits level also). I suspect there is something I'm missing here.

**Comment by guillaumekln** [comment]
There are often subtle differences in the way beam search is implemented by different frameworks. Here with the options you set, CTranslate2 and Transformers have almost the same behavior, but there is 1 difference in the code that explains this output difference.

Transformers applies the token penalties (e.g. `no_repeat_ngram_size`) after computing the softmax:

https://github.com/huggingface/transformers/blob/da971b2271728addaac5b8fd3fb68a2894e71995/src/transformers/generation/utils.py#L3040

while CTranslate2 applies the penalty before the softmax:

https://github.com/OpenNMT/CTranslate2/blob/2203ad5c8baf878a2d08e73095421e2ba033c89c/src/decoding.cc#L489-L496

Both approach can be considered correct. However, "logits" usually refers to the model output before the softmax so the CTranslate2 implementation is more consistent with the naming here.

In Transformers, if you call `logits_processor` before the softmax you get the same output as CTranslate2.

**Comment by koren-v** [comment]
Hi, I faced the same issue. However, when read @guillaumekln response, I thought that if I didn't use any special generation parameters (like `no_repeat_ngram_size`), I would get the same result. Unfortunately, seems that there are other differences in beam search implementation - or do I miss something?

To reproduce:

package versions: `transformers==4.34.0`, `ctranslate2==3.20.0` (those which were mentioned by @guillaumekln)


1. convert model
```
ct2-transformers-converter --model "google/flan-t5-base" --output_dir "ct2-t5-base"
```
2. code snippet:
```
import torch

from transformers import T5ForConditionalGeneration, AutoTokenizer
import ctranslate2


device = torch.device("cuda")

model_name = "google/flan-t5-base"
hf_model = T5ForConditionalGeneration.from_pretrained(model_name).eval().to(device)
tokenizer = AutoTokenizer.from_pretrained(model_name)


fast_model =  ctranslate2.Translator("ct2-t5-base", device="cuda")


text = "translate English to German: physician assistants are medical providers who are licensed to diagnose and treat illness and disease and to prescribe medication"


def get_out(inp, model):
    inputs = tokenizer(inp, return_tensors="pt")
    ids = model.generate(**inputs.to(device),
                         num_beams=3,
                         min_length=0,
                         max_length=1024,
                         )
    return tokenizer.batch_decode(ids, skip_special_tokens=True)[0]


def get_out_fast(inp, model):
    source = tokenizer.encode(inp)
    source = tokenizer.convert_ids_to_tokens(source)
    results = model.translate_batch([source],
                                    beam_size=3,
                                    min_decoding_length=0,
                                    max_decoding_length=1024)
    target = results[0].hypotheses[0]
    return tokenizer.decode(tokenizer.convert_tokens_to_ids(target), skip_special_tokens=True)


res_vanilla = get_out(text, hf_model)
res_fast = get_out_fast(text, fast_model)


print("Vanilla output:", res_vanilla)
print("Ctranslate output:", res_fast)
```

Output:
```
Vanilla output: physician assistants sind medical providers, die zu Diagnose und Behandlung von Krankheiten und Krankheiten und zu Verknüpfen von Medikamenten zu ermitteln.
Ctranslate output: physician assistants sind medical providers, die zu Diagnose und Behandlung von Krankheiten und Krankheiten und zu Verknüpfen von Medikamenten zu kaufen sind.
```
